src/main/java/io/kestra/plugin/openai/ChatCompletion.java

```diff
@@ -210,6 +232,32 @@ public static class Output implements io.kestra.core.models.tasks.Output {
         private Usage usage;
     }
 
+    @Getter
+    public static class PluginChatFunction extends ChatFunction {
```

Comment:
Is this class needed? Can't you do the same as with ChatMessage?

Reply:
Unfortunately, `ChatMessage` only lets you set the `parameters` type as a `Class` - it's implemented as 

```
@JsonProperty("parameters")
    private Class<?> parametersClass;
```

`parameters` is a JSON schema, so the way `ChatMessage` is written there's no way to pass through the values from the Kestra flow at runtime. from This ties in the the mixin above. Using `PluginChatFunction` alone doesn't work, because adds its own mixin here: https://github.com/TheoKanning/openai-java/blob/442cf130f6fbedbdcb70ce9d96ce572faf54c9e9/service/src/main/java/com/theokanning/openai/service/OpenAiService.java#L423

The only way I was able to override it was creating `PluginChatFunction` to provide the ability to set the `parameters` JSON schema at runtime, and then add a new mixin to prevent Jackson from falling back on the underlying `ChatFunctionCallMixIn`. 

This is all far uglier than I'd like. When I looked at the library's code a few weeks ago, it looked like he'd designed `ChatFunction` to work the way it does, and offered the ability to override things the way I did for anyone who needed different behavior. But perhaps it is worth asking him. :)

Reply:
Unfortunately, `ChatMessage` only lets you set the `parameters` type as a `Class` - it's implemented as 

```
@JsonProperty("parameters")
    private Class<?> parametersClass;
```

`parameters` is a JSON schema, so the way `ChatMessage` is written there's no way to pass through the values from the Kestra flow at runtime. from This ties in the the mixin above. Using `PluginChatFunction` alone doesn't work, because adds its own mixin here: https://github.com/TheoKanning/openai-java/blob/442cf130f6fbedbdcb70ce9d96ce572faf54c9e9/service/src/main/java/com/theokanning/openai/service/OpenAiService.java#L423

The only way I was able to override it was creating `PluginChatFunction` to provide the ability to set the `parameters` JSON schema at runtime, and then add a new mixin to prevent Jackson from falling back on the underlying `ChatFunctionCallMixIn`. 

This is all far uglier than I'd like. When I looked at the library's code a few weeks ago, it looked like he'd designed `ChatFunction` to work the way it does, and offered the ability to override things the way I did for anyone who needed different behavior. But perhaps it is worth asking him. :)

Reply:
I changed this so it no longer extends ChatFunction. The library adds a class called `ChatFunctionDynamic` that enables function calling without needing the workarounds I used before, but when I try to use 'ChatFunctionDynamic` directly, Kestra says: 

```
Cannot construct instance of `com.theokanning.openai.completion.chat.ChatFunctionDynamic` (although at least one Creator exists): cannot deserialize from Object value (no delegate- or property-based Creator)
```

Creating classes for functions and parameters solved this and also made it possible to add `Schema` annotations to provide helpful guidance in the Kestra editor.